File: Advent of Code/2024/6/solution.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <set>

struct Direction {
    int dx, dy;
};

constexpr Direction dirs[4] = {
    { 0, -1 }, { 1, 0 }, { 0, 1 }, { -1, 0 }
};

struct Field
{
    std::vector<char> cell;
    size_t width = 0;
    size_t height = 0;
    size_t start_x = 0;
    size_t start_y = 0;

    char& at(size_t x, size_t y) noexcept { 
        return cell[y * width + x];
    }

    bool addRow(const std::string& row) noexcept {
        if (width == 0)
            width = row.size();
        else if (width != row.size())
            return false;
        for (size_t i = 0, len = row.size(); i < len; ++i) {
            if (row[i] == '^') {
                start_x = i;
                start_y = height;
                break;
            }
        }
        cell.insert(cell.end(), row.begin(), row.end());
        ++height;
        return true;
    }
};

struct Solutions
{
    size_t sol_1;
    size_t sol_2;
};

Solutions solve_part_1_and_2(Field& field);

int main()
{
    Field field;
    
    {
        std::ifstream file("input.txt", std::fstream::in);
        std::string line;
        while (!file.eof()) {        
            std::getline(file, line);
            if (!field.addRow(line)) {
                std::cout << "Something went wrong!!" << std::endl;
                exit(1);
            }
        }
    }
    
    Solutions s = solve_part_1_and_2(field);
    std::cout << "Answer part 1: " << s.sol_1 << std::endl;
    std::cout << "Answer part 2: " << s.sol_2 << std::endl;

    return 0;
}

using VisitPos = std::tuple<int16_t, int16_t>;

std::set<VisitPos> just_visit(Field& field)
{
    std::set<VisitPos> visited;
    int16_t angle = 0;
    int16_t x = field.start_x;
    int16_t y = field.start_y;
    Direction dir = dirs[angle];

    while (true) 
    {
        visited.emplace(x, y);

        int nx = x + dir.dx; 
        int ny = y + dir.dy;

        if (!(0 <= nx && nx < field.width) || !(0 <= ny && ny < field.height))
            break;

        if (field.at(nx, ny) == '#') 
        {
            angle = (angle + 1) % 4;
            dir = dirs[angle];
        } 
        else 
        {
            x = nx;
            y = ny;
        }
    }   

    return std::move(visited);
}

using VisitPosAngle = std::tuple<int16_t, int16_t, int8_t>;

static size_t iters = 0;

bool detect_loop(Field& field) 
{
    std::set<VisitPosAngle> visited;
    int16_t angle = 0;
    int16_t x = field.start_x;
    int16_t y = field.start_y;
    Direction dir = dirs[angle];

    while (true) 
    {
        ++iters;

        auto st = visited.emplace(x, y, angle);
        if (!st.second)
            return true; // already visited with same angle, loop detected

        int nx = x + dir.dx; 
        int ny = y + dir.dy;

        if (!(0 <= nx && nx < field.width) || !(0 <= ny && ny < field.height))
            break;

        if (field.at(nx, ny) == '#') 
        {
            angle = (angle + 1) % 4;
            dir = dirs[angle];
        } 
        else 
        {
            x = nx;
            y = ny;
        }
    }   

    return false;
}

Solutions solve_part_1_and_2(Field& field)
{
    std::cout << "Running..." << std::endl;

    Solutions s;

    std::set<VisitPos> visited = just_visit(field);

    s.sol_1 = visited.size();
    s.sol_2 = 0;

    for (const VisitPos& v : visited)
    {
        char& c = field.at(std::get<0>(v), std::get<1>(v));
        c = '#';
        s.sol_2 += detect_loop(field);
        c = '.';
    }
    
    std::cout << "Iterations: " << iters << std::endl;

    return s;
}

File: Advent of Code/2024/6/solution.py
class Glob:
    dirs = [(0, -1), (1, 0), (0, 1), (-1, 0)]
    start_x = 0
    start_y = 0 
    size_x = 0
    size_y = 0


def main():

    with open("input.txt", "r") as file:
        lines = [ [*a.strip()] for a in file.readlines() ]

    Glob.size_y = len(lines)
    Glob.size_x = len(lines[0])

    for i, line in enumerate(lines):
        for j, c in enumerate(line):
            if c == "^":
                Glob.start_x, Glob.start_y = j, i
                break
        else:
            continue
        break

    sol_1 = solve_part_1(lines)
    print(f"Answer part 1: {sol_1}")

    sol_2 = solve_part_2(lines)
    print(f"Answer part 2: {sol_2}")



def solve_part_1(lines):

    visited = set()
    angle = 0
    x, y = Glob.start_x, Glob.start_y
    dx, dy = Glob.dirs[angle]

    while True:
        visited.add((x, y))

        nx, ny = x + dx, y + dy

        if not (0 <= nx < Glob.size_x) or not (0 <= ny < Glob.size_y):
            break
        
        if lines[ny][nx] == "#":
            angle = (angle + 1) % len(Glob.dirs)
            dx, dy = Glob.dirs[angle]
        else:
            x, y = nx, ny

    return len(visited)


def solve_part_2(lines):

    def detect_loop() -> int:

        visited = set()
        angle = 0
        x, y = Glob.start_x, Glob.start_y
        dx, dy = Glob.dirs[angle]

        while True:   
            prev_visited_size = len(visited)
            visited.add((x, y, angle))
            if (prev_visited_size == len(visited)):
                return 1 # already visited with same angle, loop detected
            
            nx, ny = x + dx, y + dy

            if not (0 <= nx < Glob.size_x) or not (0 <= ny < Glob.size_y):
                break

            if lines[ny][nx] == "#":
                angle = (angle + 1) % len(Glob.dirs)
                dx, dy = Glob.dirs[angle]
            else:
                x, y = nx, ny

        return 0


    cumul = 0

    for i, line in enumerate(lines):
        for j, c in enumerate(line):
            if c == ".":
                lines[i][j] = "#"
                cumul += detect_loop()
                lines[i][j] = "."

    return cumul


if __name__ == "__main__":
    main()